### PR TITLE
[IMP] web, mass_mailing: make domain field foldable

### PR DESCRIPTION
**Pull request by juliusc2066**
This commit allows the domain field to become foldable such that it will display its domain in the form of facets (similar to the search bar's ones) when it is folded. Also applies this functionality directly to the mass_mailing form view.

task-3339512

### CODE REVIEW

File: addons/web/static/src/core/domain_selector/domain_selector.xml [hunk 1]
```diff
@@ -3,6 +3,9 @@
 
     <t t-name="web._DomainSelector" owl="1">
         <div class="o_domain_node o_domain_tree o_domain_selector" aria-atomic="true" t-att-class="className" t-ref="root">
+            <a t-if="props.isFoldable" t-on-click="props.fold">
```
Comment (Polymorphe57) [inline addons/web/static/src/core/domain_selector/domain_selector.xml:6]: why is it not inside domainf field?

File: addons/web/static/src/views/fields/domain/domain_field.js [hunk 2]
```diff
@@ -21,29 +23,40 @@ export class DomainField extends Component {
...
 
     setup() {
         this.orm = useService("orm");
+        this.getDomainTreeDescription = useGetDomainTreeDescription();
         this.addDialog = useOwnedDialogs();
 
         this.state = useState({
             isValid: null,
             recordCount: null,
+            folded: true,
```
Comment (Polymorphe57) [inline addons/web/static/src/views/fields/domain/domain_field.js:41]: folded: this.props.isFoldable?

folded: true gives the impression that it is always folded

File: addons/web/static/src/core/domain_selector/domain_selector.js [hunk 3]
```diff
@@ -101,7 +101,9 @@ export class DomainSelector extends Component {
         defaultLeafValue: { type: Array, optional: true },
         isDebugMode: { type: Boolean, optional: true },
         readonly: { type: Boolean, optional: true },
+        isFoldable: { type: Boolean, optional: true },
```
Comment (Polymorphe57) [inline addons/web/static/src/core/domain_selector/domain_selector.js:104]: I think we should not have these props

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 4]
```diff
@@ -942,4 +942,82 @@ QUnit.module("Fields", (hooks) => {
...
+                    <sheet>
+                        <group>
+                            <field name="foo" widget="domain" options="{'model': 'partner_type', 'foldable': true}" />
+                        </group>
+                    </sheet>
+                </form>`,
+        });
+
+        // As the domain is empty, the "Match all records" span should be visible
+        assert.containsOnce(target, ".o_field_domain .d-print-contents span");
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/fields/domain_field_tests.js:965]: the selector is not very informative

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 5]
```diff
@@ -942,4 +942,82 @@ QUnit.module("Fields", (hooks) => {
...
+
+        // Unfold the domain
+        await click(target, ".o_field_domain .align-items-center");
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/fields/domain_field_tests.js:968]: the selector is not very informative

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 6]
```diff
@@ -942,4 +942,82 @@ QUnit.module("Fields", (hooks) => {
...
+        await nextTick();
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/fields/domain_field_tests.js:969]: why?

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 7]
```diff
@@ -942,4 +942,82 @@ QUnit.module("Fields", (hooks) => {
...
+
+        // As the domain is empty, there should be a button to add the first
+        // domain part
+        assert.containsOnce(target, ".o_domain_add_first_node_button");
+
+        // Clicking on the button should add the [["id", "=", "1"]] domain, so
+        // there should be a field selector in the DOM
+        await click(target, ".o_domain_add_first_node_button");
+        assert.containsOnce(target, ".o_model_field_selector", "there should be a field selector");
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/fields/domain_field_tests.js:978]: avoid message in the asserts

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 8]
```diff
@@ -942,4 +942,82 @@ QUnit.module("Fields", (hooks) => {
...
+        assert.strictEqual(
+            document.body.querySelector(".o_model_field_selector_popover_item_name").textContent,
+            "Color index"
+        );
+
+        // Clicking on this field should close the popover, then changing the
+        // associated value should reveal one matched record
+        await click(document.body.querySelector(".o_model_field_selector_popover_item_name"));
+
+        const input = target.querySelector(".o_domain_leaf_value_input");
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/fields/domain_field_tests.js:997]: use editInput?

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 9]
```diff
@@ -942,4 +942,82 @@ QUnit.module("Fields", (hooks) => {
...
+        );
+
+        // Saving the form view should show a readonly domain containing the
+        // "color" field
+        await clickSave(target);
+        assert.ok(target.querySelector(".o_field_domain").textContent.includes("Color index"));
+
+        // Fold domain selector
+        await click(target, ".o_domain_selector a i");
+        await nextTick();
```
Comment (Polymorphe57) [inline addons/web/static/tests/views/fields/domain_field_tests.js:1014]: why?

File: addons/web/static/src/views/fields/domain/domain_field.xml [hunk 10]
```diff
@@ -5,45 +5,77 @@
...
-                            </button>
+                <t t-if="props.isFoldable and state.folded">
+                    <div class="d-flex align-items-center" t-on-click="() => state.folded = false">
+                        <i data-tooltip="Modify filter" class="fa fa-lg fa-caret-right pe-2"/>
+                        <div class="d-print-contents">
+                            <t t-if="!state.facets.length">
+                                <span>Match <strong>all records</strong></span>
+                            </t>
+                            <t t-foreach="state.facets" t-as="facet" t-key="facet_index">
+                                <div class="d-inline-flex align-items-stretch text-nowrap px-1 position-relative o_facet_with_domain"
```
Comment (Polymorphe57) [inline addons/web/static/src/views/fields/domain/domain_field.xml:16]: the class o_facet_with_domain is not necessary here I think.
Please, review the classes and be sure that only the necessary classes are set

File: addons/web/static/tests/views/fields/domain_field_tests.js (hunk 5 above)
Comment (juliusc2066) [reply addons/web/static/tests/views/fields/domain_field_tests.js:968]: I don't have much more to work with here

File: addons/web/static/tests/views/fields/domain_field_tests.js (hunk 6 above)
Comment (juliusc2066) [reply addons/web/static/tests/views/fields/domain_field_tests.js:969]: No point indeed

**Comment by Polymorphe57** [comment]
robodoo r+

**Pull request closed without merging**